core/field.ts

```diff
@@ -1046,12 +1085,16 @@ export abstract class Field<T = any> implements IASTNodeLocationSvg,
    * @param opt_newValue The value to be validated.
    * @returns The validated value, same as input by default.
    */
-  protected doClassValidation_(opt_newValue?: AnyDuringMigration):
-      AnyDuringMigration {
+  protected doClassValidation_(opt_newValue?: U): ClassValidation<T, U> {
+    // NOTE: Although `undefined` is never returned here, it is valid for
```

Comment:
Good point - I can see how my original wording and placement is misleading. I'll go ahead and move the note to the tsdoc and reword it to encompass the `undefined` rule.

Thinking through the rule itself, it can actually be enforced with simple function overloading. (Creating a union type for the rule would be complex and require a slight implementation change.) I'll go ahead and upload that implementation, though let me know what you think!